engines/router/missionctl/mission_control_upi_test.go

```diff
@@ -0,0 +1,255 @@
+			},
+			mockReturn: fiber.NewResponseQueueFromResponses(
+				fiber.NewErrorResponse(
+					fiberErrors.FiberError{
+						Code:    13,
+						Message: "fiber: request cannot be completed: err",
+					})),
+		},
+		{
+			name: "error non proto payload",
+			expectedErr: &errors.TuringError{
+				Code:    14,
+				Message: "unable to parse fiber response into grpc response",
+			},
+			mockReturn: fiber.NewResponseQueueFromResponses(fiberhttp.NewHTTPResponse(
+				&http.Response{
+					StatusCode: http.StatusOK,
+					Body:       ioutil.NopCloser(bytes.NewReader([]byte("dummy res"))),
+				})),
+		},
+		{
+			name: "error wrong response payload type",
+			expectedErr: &errors.TuringError{
+				Code:    14,
+				Message: "unable to unmarshal into expected response proto",
+			},
+			mockReturn: fiber.NewResponseQueueFromResponses(&fibergrpc.Response{
+				Message: &upiv1.ResponseMetadata{
+					PredictionId: "123",
+				},
+			}),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			mockFiberRouter := &mocks.Component{}
+			mockFiberRouter.On("Dispatch", mock.Anything, mock.Anything).Return(tt.mockReturn, nil)
+			mc := missionControlUpi{fiberRouter: mockFiberRouter}
+
+			got, err := mc.Route(context.Background(), &fibergrpc.Request{})
+			if tt.expectedErr != nil {
+				require.Equal(t, tt.expectedErr, err)
+			} else {
+				require.True(t, compareUpiResponse(got, tt.expected), "response not equal to expected")
+			}
+
+		})
+	}
+}
+
+func compareUpiResponse(x *upiv1.PredictValuesResponse, y *upiv1.PredictValuesResponse) bool {
+	return cmp.Equal(x, y,
+		cmpopts.IgnoreUnexported(
+			upiv1.PredictValuesResponse{},
+			upiv1.Table{},
+			upiv1.NamedValue{},
+			upiv1.ResponseMetadata{},
+		))
+}
+
+func benchmarkGrpcRoute(payloadFileName string, b *testing.B) {
+
+	mc, err := NewMissionControlUPI(benchmarkConfig, false)
+	require.NoError(b, err)
+
+	upiRequest := &upiv1.PredictValuesRequest{}
+	fileByte, err := ioutil.ReadFile(filepath.Join("testdata", payloadFileName))
+	require.NoError(b, err)
+	err = protojson.Unmarshal(fileByte, upiRequest)
+	require.NoError(b, err)
+
+	req := &fibergrpc.Request{
+		Message: upiRequest,
+	}
+
+	b.ResetTimer()
+	for n := 0; n < b.N; n++ {
+		benchMarkUpiResp, benchMarkUpiErr = mc.Route(context.Background(), req)
+	}
+}
+
+func benchmarkPlainGrpc(payloadFileName string, b *testing.B) {
+
+	upiRequest := &upiv1.PredictValuesRequest{}
+	fileByte, err := ioutil.ReadFile(filepath.Join("testdata", payloadFileName))
+	require.NoError(b, err)
+	err = protojson.Unmarshal(fileByte, upiRequest)
+	require.NoError(b, err)
+
+	conn, err := grpc.Dial(fmt.Sprintf(":%d", grpcport1), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	require.NoError(b, err)
+	client := upiv1.NewUniversalPredictionServiceClient(conn)
+
+	b.ResetTimer()
+	for n := 0; n < b.N; n++ {
+		benchMarkUpiResp, _ = client.PredictValues(context.Background(), upiRequest)
```

Comment:
Can you make the response from server adjustable? Currently it only returns a mocked response which is not representative of the actual use case. One alternative that you could take is to populate the `prediction_result_table` with the `prediction_table` of the request.

Reply:
Additionally, can you try to profile the code (using pprof) when executing the benchmark to see whether there is some low hanging fruit for optimization. I would love to reduce the performance to be as close as the plain grpc one.